### PR TITLE
Add Error Handling to the Signup and User Dashboard pipeline

### PR DESCRIPTION
**Pull request by Obinnabii**
- Add ```UserServlet``` Error handling to match the ```OpenSesame``` Error Handling conventions. We are transitioning to using the ```ErrorResponse``` module to send errors from servlet requests.
- Use the ```standardizedFetchErrorHandler``` function for for catching and dealing with errors recieved from the ```UserServlet``` in the ```user.js``` module.

### CODE REVIEW

File: src/main/webapp/user.js [hunk 1]
```diff
@@ -6,38 +11,40 @@
...
+
+  return errorFormattedFetchRequest.then((response) => response.json())
+    .then((user) => {
+      return user;
+    })
 }
 
 /**
  * Posts a user to the servlet
  * @param {URLSearchParams} userParams parameters with the users information
```
Comment (Obinnabii) [inline src/main/webapp/user.js:30]:
```suggestion
 * @param {URLSearchParams} userParams parameters with the users information
 * @return {Promise} response from '/user' POST
```

File: src/main/webapp/signup.js [hunk 2]
```diff
@@ -4,10 +4,7 @@
  * @property {string[]} interestTags
  */
 import {gitHubAuthorizer} from './authorization.js';
-import {
-  standardizeFetchErrors,
-  makeRelativeUrlAbsolute,
-} from './js/fetch_handler.js';
+import {postUser} from '.user.js';
```
Comment (Obinnabii) [inline src/main/webapp/signup.js:7]:
```suggestion
import {postUser} from './user.js';
```

File: src/main/java/com/google/opensesame/servlets/UserServlet.java [hunk 3]
```diff
@@ -55,16 +72,29 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String userGitHubID = request.getParameter("githubID");
-    if (userGitHubID.trim().isEmpty()) {
-      response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Invalid ID");
-      return;
-    }
-    ArrayList<String> userTags = new ArrayList<>(Arrays.asList(request.getParameterValues("tags")));
+    try {
+      String userGitHubID = request.getParameter("githubID");
+      if (userGitHubID.trim().isEmpty()) {
```
Comment (Richie78321) [inline src/main/java/com/google/opensesame/servlets/UserServlet.java:77]: `isEmpty()` is Java 11 only. Consider:
```suggestion
      String userGitHubID = request.getParameter("githubID").trim();
      if (userGitHubID == null || userGitHubID.length() == 0) {
```

https://stackoverflow.com/questions/32567208/best-way-to-verify-string-is-empty-or-null

File: src/main/webapp/signup.js [hunk 4]
```diff
@@ -116,21 +113,7 @@ function createSignupRequest() {
...
-  });
-
-  const errorFormattedFetchRequest = standardizeFetchErrors(
-      fetchRequest,
-      'Failed to communicate with the server. Please try again later.',
-      'An error occcured while creating your account.' +
-          ' Please try again later.');
-
-  return errorFormattedFetchRequest;
+  return postUser(encodedBody);
```
Comment (Richie78321) [inline src/main/webapp/signup.js:116]: Is there a reason why this code was moved? Will there ever be another use beyond signup where a user will need to be posted to the `/users` servlet? If your intention here is to make the post functionality more generally accessible if we intend to allow users to edit their own profiles in the future, consider following [RESTful API standards](https://www.restapitutorial.com/lessons/httpmethods.html#:~:text=The%20primary%20or%20most%2Dcommonly,but%20are%20utilized%20less%20frequently.) and using a [`doPut()`](https://tomcat.apache.org/tomcat-5.5-doc/servletapi/javax/servlet/http/HttpServlet.html#doPut(javax.servlet.http.HttpServletRequest,%20javax.servlet.http.HttpServletResponse)) and sending a PUT request in that case. That will allow us to keep the two functionalities separate and follow proper standards.

File: src/main/webapp/user.js [hunk 5]
```diff
@@ -6,38 +11,41 @@
...
-      .then(errorHandling).then((response) => response.json())
+  const fetchRequest = fetch(makeRelativeUrlAbsolute('/user?githubID=' + user));
+
+  const errorFormattedFetchRequest = standardizeFetchErrors(
+      fetchRequest,
+      'Failed to communicate with the server. Please try again later.',
+      'An error occcured while retrieving this account.' +
+    ' Please try again later.');
+
+  return errorFormattedFetchRequest.then((response) => response.json())
```
Comment (Richie78321) [inline src/main/webapp/user.js:22]: Nice!

File: src/main/java/com/google/opensesame/servlets/UserServlet.java [hunk 6]
```diff
@@ -31,19 +32,35 @@ public void init() {
...
+          "The user requested could not be found on the server."
+              + "Please ensure that the user has an account with us.");
+      return;
+    }
     PersonObject userObject;
     try {
-      userObject = new PersonBuilder().buildPersonObject(datastore.get(userKey));
-      System.out.println(userObject);
+      userObject = new PersonBuilder().buildPersonObject(userEntity);
     } catch (Exception e) {
```
Comment (Richie78321) [inline src/main/java/com/google/opensesame/servlets/UserServlet.java:55]: This looks good, although I'm wondering if we could move this functionality to a public static function like `getPersonObjectById(id)` that we could use in other places like the `hasProfile` check in the auth servlet?

File: src/main/java/com/google/opensesame/servlets/UserServlet.java [hunk 7]
```diff
@@ -31,19 +32,35 @@ public void init() {
   }
 
   @Override
-  // Get a specific user. Return null if not found. TODO: User Validation
+  // Get a specific user. Return null if not found.
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
     String userGithub = request.getParameter("githubID");
```
[the remark this answers is not in the document]
Comment (Obinnabii) [reply src/main/java/com/google/opensesame/servlets/UserServlet.java:37]: will do by the next PR

File: src/main/webapp/signup.js (hunk 4 above)
Comment (Obinnabii) [reply src/main/webapp/signup.js:116]: Oh good catch! I'll use the ```doPut()```. My reason for moving the function was to make it more accessible but I guess it won't really be used anywhere else. That being said, I'd still like to have it in the ```user.js``` module. I like having all the functions that use the ```/user``` servlet in the same place. If you think it's better in another module, that's fine too

File: src/main/java/com/google/opensesame/servlets/UserServlet.java [hunk 8]
```diff
@@ -55,16 +72,29 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+      Entity personEntity =
+          new PersonBuilder().gitHubID(userGitHubID).interestTags(userTags).buildPersonEntity();
+      datastore.put(personEntity);
+    } catch (Exception e) {
+      ErrorResponse.sendJsonError(
+          response,
+          e.getMessage() + "/n/n" + e.getStackTrace(),
+          HttpServletResponse.SC_BAD_REQUEST,
+          "User informatiuon could not be sent to datastore");
+    }
```
[the remark this answers is not in the document]
Comment (Obinnabii) [reply src/main/java/com/google/opensesame/servlets/UserServlet.java:97]: Interesting point! do we want to expose really specific error messages to the user? I'll add more try blocks. As it is, when an error occurs, we get a specific error message with an exposed callstack returned to the client but the user gets a vague ```"user information could not be sent to datastore"``` alert

File: src/main/java/com/google/opensesame/servlets/UserServlet.java (hunk 6 above)
Comment (Obinnabii) [reply src/main/java/com/google/opensesame/servlets/UserServlet.java:55]: will do!

File: src/main/java/com/google/opensesame/servlets/UserServlet.java [hunk 9]
```diff
@@ -31,19 +32,35 @@ public void init() {
...
 
   @Override
-  // Get a specific user. Return null if not found. TODO: User Validation
+  // Get a specific user. Return null if not found.
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
     String userGithub = request.getParameter("githubID");
     Key userKey = KeyFactory.createKey(PersonBuilder.ENTITY_NAME, userGithub);
-    System.out.println(userKey);
 
+    Entity userEntity;
```
Comment (Obinnabii) [inline src/main/java/com/google/opensesame/servlets/UserServlet.java:40]: > This looks good, although I'm wondering if we could move this functionality to a public static function like `getPersonObjectById(id)` that we could use in other places like the `hasProfile` check in the auth servlet?
```suggestion
    Entity userEntity; // TODO: Abstract doGet() into it's own function 
```

File: src/main/java/com/google/opensesame/servlets/UserServlet.java [hunk 10]
```diff
@@ -55,16 +72,29 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+      if (userGitHubID == null || userGitHubID.trim().length() == 0) {
+        ErrorResponse.sendJsonError(
+            response,
+            "Empty GitHub ID",
+            HttpServletResponse.SC_BAD_REQUEST,
+            "There is no GitHub ID affiliated with this request");
+        return;
+      }
+      ArrayList<String> userTags =
+          new ArrayList<>(Arrays.asList(request.getParameterValues("tags")));
```
Comment (billyeh) [inline src/main/java/com/google/opensesame/servlets/UserServlet.java:86]: Does this part need to be in the `try` block? Best to keep the scope of it as limited as possible and move these lines out if not.

File: src/main/java/com/google/opensesame/servlets/UserServlet.java [hunk 11]
```diff
@@ -31,19 +32,35 @@ public void init() {
...
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String userGithub = request.getParameter("githubID");
+    String userGithub = request.getParameter("githubID"); // TODO:Check for null values
     Key userKey = KeyFactory.createKey(PersonBuilder.ENTITY_NAME, userGithub);
-    System.out.println(userKey);
 
+    Entity userEntity; // TODO: Abstract doGet() into it's own function
+    try {
+      userEntity = datastore.get(userKey);
+    } catch (Exception e) {
```
Comment (billyeh) [inline src/main/java/com/google/opensesame/servlets/UserServlet.java:43]: Is it possible to be a little more specific about what Exceptions we expect to be thrown here? For example, it should just be EntityNotFoundException here. https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/EntityNotFoundException.html

And below in the other try/catches.

File: src/main/webapp/signup.js (hunk 4 above)
Comment (Richie78321) [reply src/main/webapp/signup.js:116]: As long as we're on the same page that it really will only be used for signup, I'd prefer if it stayed in the original file `signup.js`. That way we can avoid confusion with the post request vs. put request and we can have more freedom with changes to this function down the line. Also, when you moved it, you exposed a lot of the error checks and content limiting. Before one was only able to make a post request with specific signup body data, and now a post request can be made with any URLEncoded data.

Comment (Obinnabii) [inline src/main/webapp/signup.js:116]:
```suggestion
 const fetchRequest = fetch(makeRelativeUrlAbsolute('/user'), {
      method: 'POST',	
    headers: {	
      'Content-Type': 'application/x-www-form-urlencoded',	
    },	
    body: encodedBody,	
  });	

  const errorFormattedFetchRequest = standardizeFetchErrors(	
      fetchRequest,	
      'Failed to communicate with the server. Please try again later.',	
      'An error occcured while creating your account.' +	
          ' Please try again later.');	

  return errorFormattedFetchRequest;
```

**Comment by Obinnabii** [comment]
I am going to overhaul this User Servlet implementation entirely with Objectify

**Pull request merged**